### PR TITLE
fix: Ignore shadowRoots on elements that don't allow them

### LINKED ISSUE #505: Ignore shadowRoots on elements that don't allow them
Came across an issue of axe 3.0 not being able to build its tree correctly. Turned out it was caused by Chrome, which has recently added a shadow root to every HTML element. This tripped `axe.utils.flattenTree` up massively.

The solution is to only build the shadow tree for elements allowed to have shadow trees:
https://dom.spec.whatwg.org/#dom-element-attachshadow

Only valid custom element names, "article", "aside", "blockquote", "body", "div", "footer", "h1", "h2", "h3", "h4", "h5", "h6", "header", "main" "nav", "p", "section", or "span",

### PR DESCRIPTION
**Pull request by WilcoFiers**
Closes #505

### CODE REVIEW

File: lib/core/utils/flattened-tree.js [hunk 1]
```diff
@@ -51,6 +51,20 @@ function getSlotChildren(node) {
 	return retVal;
 }
 
+const possibleShadowRoots = ['article', 'aside', 'blockquote',
```
Comment (marcysutton) [inline lib/core/utils/flattened-tree.js]: What about custom elements? Could we do a blacklist approach instead of a whitelist and exclude `html`?

Comment (WilcoFiers) [reply lib/core/utils/flattened-tree.js]: Custom elements are allowed as long as they have a dash in the name. Chrome doesn't let you attach shadow trees to other stuff, so I think this works best.

**Comment by dylanb** [comment]
@wilco what about v0?

**Comment by marcysutton** [comment]
@dylanb the only API I'd wonder about is `.shadowRoot`, which I see referenced in this v0 gist from Eric Bidel: https://gist.github.com/ebidel/65b4565947c6399c82ac6f7d1a276442

I can't find a spec with a list of allowed elements, so with custom elements allowed I'd assume this is fine.

**Comment by marcysutton** [comment]
I'm going to merge this, because we need to release an alpha.2 for aXe coconut.

File: lib/core/utils/flattened-tree.js (hunk 1 above)
Comment (dylanb) [reply lib/core/utils/flattened-tree.js]: @WilcoFiers could you add a test to make sure that that statement does not change?

Comment (WilcoFiers) [reply lib/core/utils/flattened-tree.js]: You mean test that something in the codebase didn't modify this array somehow? I could freeze it if you want.

Comment (marcysutton) [reply lib/core/utils/flattened-tree.js]: Maybe write a test for custom elements?

Comment (dylanb) [reply lib/core/utils/flattened-tree.js]: @WilcoFiers No, what I was suggesting is that we write a test to make sure you cannot attach a shadow root to a custom element without a hyphen

**Pull request merged**